### PR TITLE
[WIP] Fix #11978: Filter out gdb_io requests when they are out of sections if sections are available

### LINKED ISSUE #11978: esil analysis breaks gdb connection with mips32 binaries
test done on a fedora 28
```
$ r2 -v
radare2 3.1.0-git 19911 @ linux-x86-64 git.3.0.1-70-g650628d0f
commit: 650628d0f6b35d993fa3d07c80ff902d36c91daf build: 2018-10-27__15:14:05
```

I'm trying to debug a mips32 binary with qemu's gdb stubs.
```
r2 -e anal.gp=loc._gp -e anal.gp2=loc._gp -e bin.baddr=0x400000 -e dbg.exe.path=$PWD/example -e cfg.bigendian=true -a mips -b 32 -d gdb://localhost:1234
```
```
file example
example: ELF 32-bit MSB executable, MIPS, MIPS32 rel2 version 1 (SYSV), dynamically linked, interpreter /lib/ld.so.1, for GNU/Linux 3.2.0, BuildID[sha1]=449e7f2913faba0676cbf7c8d87ca63fcf993b64, not stripped
```
seems like when disassembling, esil analyses breaks the connection, the unique workaround I have is setting 
```
e emu.str=false
```

But this is annoying because without a debugger attached, esil is able to resolve things like:
```
0x004008f4      3c020040       lui v0, 0x40                ; segment.ehdr                           
0x004008f8      24440bf0       addiu a0, v0, 0xbf0         ; "\nAccess denied.\n" str.Access_denied.
```

when setting a breakpoint in `read_packet` at 
```
eprintf ("%s: read failed\n", __func__);
```
I get this stacktrace:

```
Breakpoint 1, read_packet (g=0x55555596d580, vcont=false) at src/packet.c:172
172				eprintf ("%s: read failed\n", __func__);
Missing separate debuginfos, use: dnf debuginfo-install python2-libs-2.7.15-4.fc28.x86_64
(gdb) bt
#0  read_packet (g=0x55555596d580, vcont=false) at src/packet.c:172
#1  0x00007ffff5911a76 in gdbr_read_memory_page (g=0x55555596d580, address=4294967295, buf=0x7fffffffc4d0 '\377' <repeats 63 times>, len=1) at src/gdbclient/core.c:562
#2  0x00007ffff5911b8d in gdbr_read_memory (g=0x55555596d580, address=4294967295, buf=0x7fffffffc4d0 '\377' <repeats 63 times>, len=63) at src/gdbclient/core.c:587
#3  0x00007ffff58de5d6 in debug_gdb_read_at (buf=0x7fffffffc4d0 '\377' <repeats 63 times>, sz=63, addr=4294967295) at p/io_gdb.c:30
#4  0x00007ffff58deba4 in __read (io=0x555555812ca0, fd=0x55555589e7a0, buf=0x7fffffffc4d0 '\377' <repeats 63 times>, count=63) at p/io_gdb.c:173
#5  0x00007ffff58f169b in r_io_plugin_read (desc=0x55555589e7a0, buf=0x7fffffffc4d0 '\377' <repeats 63 times>, len=63) at plugin.c:135
#6  0x00007ffff58f4d04 in r_io_desc_read (desc=0x55555589e7a0, buf=0x7fffffffc4d0 '\377' <repeats 63 times>, len=63) at desc.c:194
#7  0x00007ffff58f5354 in r_io_desc_read_at (desc=0x55555589e7a0, addr=4294967295, buf=0x7fffffffc4d0 '\377' <repeats 63 times>, len=63) at desc.c:326
#8  0x00007ffff58efefa in r_io_pread_at (io=0x555555812ca0, paddr=4294967295, buf=0x7fffffffc4d0 '\377' <repeats 63 times>, len=63) at io.c:286
#9  0x00007ffff58f00cb in r_io_read_at (io=0x555555812ca0, addr=4294967295, buf=0x7fffffffc4d0 '\377' <repeats 63 times>, len=63) at io.c:333
#10 0x00007ffff791e622 in myregwrite (esil=0x555555a4b390, name=0x5555559f58b0 "gp", val=0x7fffffffc598) at disasm.c:3954
#11 0x00007ffff614e5c9 in r_anal_esil_reg_write (esil=0x555555a4b390, dst=0x5555559f58b0 "gp", num=4325376) at esil.c:656
#12 0x00007ffff614e90c in esil_eq (esil=0x555555a4b390) at esil.c:721
#13 0x00007ffff615450f in runword (esil=0x555555a4b390, word=0x7fffffffc670 "=") at esil.c:2845
#14 0x00007ffff6154a0d in r_anal_esil_parse (esil=0x555555a4b390, str=0x555555a4c3b5 "") at esil.c:2989
#15 0x00007ffff791f689 in ds_print_esil_anal (ds=0x555555a4b730) at disasm.c:4251
#16 0x00007ffff79222d9 in r_core_print_disasm (p=0x5555557c3620, core=0x55555575e5a0 <r>, addr=4196448, buf=0x5555559043d0 "'\275\377X\257\277", len=80, l=5, invbreak=0, cbytes=0,
    json=false, pdf=0x0) at disasm.c:4976
#17 0x00007ffff788a112 in cmd_print (data=0x55555575e5a0 <r>, input=0x555555a14601 "d 5") at cmd_print.c:4714
#18 0x00007ffff78f0209 in r_cmd_call (cmd=0x555555827110, input=0x555555a14600 "pd 5") at cmd_api.c:239
#19 0x00007ffff78a8999 in r_core_cmd_subst_i (core=0x55555575e5a0 <r>, cmd=0x555555a14600 "pd 5", colon=0x0, tmpseek=0x7fffffffd5ff) at cmd.c:2929
#20 0x00007ffff78a585e in r_core_cmd_subst (core=0x55555575e5a0 <r>, cmd=0x555555a14600 "pd 5") at cmd.c:1937
#21 0x00007ffff78aae8f in r_core_cmd (core=0x55555575e5a0 <r>, cstr=0x555555a1c970 "pd 5", log=1) at cmd.c:3650
#22 0x00007ffff7819161 in r_core_prompt_exec (r=0x55555575e5a0 <r>) at core.c:2656
#23 0x000055555555b8f1 in main (argc=17, argv=0x7fffffffdb18, envp=0x7fffffffdba8) at radare2.c:1441

```
going a few frame up:
```
#14 0x00007ffff6154a0d in r_anal_esil_parse (esil=0x555555a4b390, str=0x555555a4c3b5 "")
    at esil.c:2989
2989			if (!runword (esil, word)) {
(gdb) p str - 13
$1 = 0x555555a4c3a8 "0x420000,gp,="

```

seems like esil is just trying to set the gp register to some value.
But the problem is that this value is not a valid address.
but myregwrite is still performing a memory read on it:

```
#9  0x00007ffff58f00cb in r_io_read_at (io=0x555555812ca0, addr=4294967295, buf=0x7fffffffc4d0 '\377' <repeats 63 times>, len=63) at io.c:333
#10 0x00007ffff791e622 in myregwrite (esil=0x555555a4b390, name=0x5555559f58b0 "gp", val=0x7fffffffc598) at disasm.c:3954
```

4294967295 is 0xffffffffffffffff

this value comes from the another call to `r_io_read_at` in myregwrite (disasm.c:3932): 
```
             (void)r_io_read_at (esil->anal->iob.io, *val, (ut8*)str, sizeof (str)-1);
```
checking val in dbg:
```
(gdb) print /x *val
$5 = 0x420000
```

so esil is trying to access addresses (in order to find strings) where it should not.
some basic instructions like this one will make the connection crash because of this:
```
0x00400870      3c1c0042       lui gp, 0x42         
0x00400874      279c9010       addiu gp, gp, -0x6ff0
```
just because 0x420000 does not exist. 
Using `r_io_is_valid_offset` everywhere in `myregwrite` wasn't better.

### PR DESCRIPTION
**Pull request by kamou**
in this first patch I do the filtering out.
But I need to add commands in order to allow hand crafting sections
this PR is related to #11978

### CODE REVIEW

File: libr/io/p/io_gdb.c [hunk 1]
```diff
@@ -8,6 +8,7 @@
 #define IRAPI static inline
 #include <libgdbr.h>
 #include <gdbclient/commands.h>
+#include <r_core.h>
```
Comment (radare) [inline libr/io/p/io_gdb.c]: you cant do this, rcore depends on rio. rio CANT depend on rcore.

use rcorebind like i have explained, if not i can do this change and y ou can base your pr on top of this

File: libr/io/p/io_gdb.c [hunk 2]
```diff
@@ -170,6 +172,20 @@ static int __read(RIO *io, RIODesc *fd, ut8 *buf, int count) {
 	if (!desc || !desc->data) {
 		return -1;
 	}
+
+	RCore *core = io->user;
+	RList *sections = r_bin_get_sections (core->bin);
```
Comment (radare) [inline libr/io/p/io_gdb.c]: you cant do t his either, rbin depends on  rio not the other way  around

File: libr/io/p/io_gdb.c (hunk 1 above)
Comment (kamou) [reply libr/io/p/io_gdb.c]: I'm not sure to understand what you are asking for with rcorebind. Should I add a new function pointer for sections access ?

**Comment by XVilka** [comment]
@kamou any updates on this one? Would be nice to have this in 3.2.0

**Comment by ret2libc** [comment]
Closing this as it was inactive for about 3 months. Please feel free to open a new PR if you get back to this! Thanks!

**Comment by radare** [comment]
This pr can be solved if we implement the io windows thing. Anyone?

**Pull request closed without merging**